### ISSUE TITLE
Add indicator to resources that are unique

### ISSUE THREAD

**Issue by Zireael07** [godotengine/godot-proposals#904]
**Describe the project you are working on:**
N/A (generic issue)

**Describe the problem or limitation you are having in your project:**
I forget which resources are unique and which are not (easy when the project is big)
This is especially important for a) shaders and b) collision shapes (we had many reports where people duplicated area2d/area3d and wondered why changing one's shape changed the other's, too)

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Tell us which resources are unique (since all resources are shared by default)

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Add an icon to the inspector for unique resources

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Nope.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Can't be done via add-on, as this changes the editor inspector.

Original issue: https://github.com/godotengine/godot/issues/6922 (I reworded it to indicate unique resources instead of shared since all resources are shared by default)

**Comment by Cykyrios** [comment]
Additionally, it could be useful to add an entry to the resource's contextual menu to view owners, as the only way to do this that I know of is to first click on "Show in FileSystem", and then "View Owners" - which is less intuitive, but still works. An icon for unique resources would definitely help to know one doesn't need to check where the resource is being used.

**Comment by Zireael07** [comment]
Related issue: https://github.com/godotengine/godot/issues/16680

**Comment by remram44** [comment]
Blender does it like this:

![screenshot from Blender](https://user-images.githubusercontent.com/426784/83061231-a8d7b180-a02a-11ea-94ac-7ca965dc3e9b.png)

**Comment by lukostello** [comment]
What if you made something unique, but too far down the road realize it was a mistake, maybe there should be a means of making it "conform" as it were.

**Comment by lukostello** [comment]
This is probably my misunderstanding talking here. But "local_to_scene" and "make_unique" seem very similar to me. Is it possible that these functionalities could be combined into a single tool?

**Comment by Zireael07** [comment]
Another case of a person gettting confused by things being shared: https://github.com/godotengine/godot/issues/54062#issuecomment-948462426

While I understand it has to be shared by default for performance reasons, this issue really really needs improvement. Otherwise we'll keep having those sorts of issues raised every couple of months, and potentially even more when 4.0 gets more widely used.

**Comment by Calinou** [comment]
As pointed out in https://github.com/godotengine/godot-proposals/issues/5921, the number of *users* for each resource should be displayed similar to Blender.

**Comment by KoBeWi** [comment]
I checked if `get_reference_count()` could be useful for this functionality, but no:
![godot windows editor dev x86_64_1LG4nqAsmb](https://user-images.githubusercontent.com/2223172/206863210-4507415f-a14d-4de5-b683-5bf1c9d74bf8.gif)
Determining unique resources might be tricky.

**Comment by QbieShay** [comment]
There is the dependency detection in the editor, could that be used?

**Comment by KoBeWi** [comment]
Not really, it only lists scenes/resources that have a dependency using `ext_resource`. It doesn't have information on how many times the given dependency is used and it won't work with built-in resources.

**Comment by QbieShay** [comment]
But in order to be shared across scenes, it has to be an ext_resource, no? Even if it's a built-in resource of scene A, scene B will reference it as an external resource?

It doesn't fully solve the problem but at least i think that if you have an open scene, you can see immediately if a resource is shared between parts of your scene, while it's realy tricky if it's another scene.

**Comment by KoBeWi** [comment]
> But in order to be shared across scenes, it has to be an ext_resource, no?

Well, yeah, but resources shared between scenes are not really a problem. More common is that resources are shared within a single scene, e.g. you make a collision shape node, add it a shape and then duplicate the node. The shape is shared. Or you make AnimationPlayer, add animation, but then decide to split the animation in multiple objects and duplicate the AnimationPlayer. The animations are shared. Or you make a Theme (even outside any scene) and make a Button style, but then you want the same style with another color, so you make new Button type and copy the StyleBox. The StyleBoxes are shared. etc. etc.

**Comment by desastreger** [comment]
Perhaps link to:
https://github.com/godotengine/godot-proposals/issues/317

**Comment by farfalk** [comment]
Maybe the sub-resource ID (for resources in the same scene) + the ext-resource ID (for resources coming from different scenes) + the new UID system could be useful for this? In my case I understood that the same resource (material) was being used in two different places of the same scene using its sub-resource id and manually searching for it in the tscn.

I wonder if an Editor-side addon could make use of them to build at least the "unique detector" functionality

**Comment by farfalk** [comment]
As asked by Fire in the godot dev chat, status report of the Proof of Concept addon:

> The main resource (a material) is saved as a .tres and used exactly once in the whole project. The nested resource (a material in next pass) is also saved as a .tres, and it's used in the main resource (duh), in another node of the same scene of the main resource, and in another scene. 
![Image](https://github.com/user-attachments/assets/03261978-9047-44b2-8866-8f834ea29135)

What I'd like to do more:

- figure out how to build a "resource users database" in the .godot folder, so that it doesn't need to parse the whole project each time
- is it possible to show this as a symbol in the preview?
- maybe a popup that shows all the users if not unique?
- probably makes sense to show if it's marked "local to scene" as well?

Also asked passivestar for a little consultation about how it should appear

**Comment by Zireael07** [comment]
Great to see progress being made on this!

**Comment by Repiteo** [comment]
Implemented by godotengine/godot#109458
